Geofeatures Leaflet implementation

lmgonzales opened the pull request:
Provides the additional option to use Leaflet with the Geographic Features functionality for both the MapSelection and MapTab features. 

@demiankatz , there were a couple of files where I had some questions for you: 

- MapSelection.php 
-- I re-wrote much of this code, but wasn't sure how to handle the setConfig method, so I just set all the options in this method. It seems like this could be simpler, but I'm not sure how. 
- theme.config.php 
-- I included geofeatures.css in the css file list in this file, but am not 100% sure that's the best place to put it. 

Thanks!

demiankatz commented:
@lmgonzales, I'll start by answering your questions above.

1.) Regarding setConfig, that code is necessary to implement the conditional .ini loading supported by the MapSelection module, as documented in searches.ini:

<pre>
; MapSelection:[ini section]:[ini name]
;       Enable geographic searching capability via OpenLayers3 API by activating
;       this module. Records must be indexed using the geographic search and display
;       fields. See the marc_local.properties file for more information on indexing.
;       Default settings and more comments may be found in the  [MapSelection]
;       section in this file. The section name and ini file name loaded by the
;       module may be overridden through the [ini section]/[ini name] parameters.
</pre>

The question is, do we care about that functionality? The main reason to allow ini section/ini name to be overridden is so that you can load MapSelection with different sets of configurations in different contexts. If that would never happen, these settings are pointless. So, I think there are two options:

a.) If the configuration is pointless, drop it from the documentation and note in the changelog that it is gone.

b.) If the configuration is useful, make the configuration loading class more flexible, and do the loading inside setConfig based on the incoming settings.

2.) Regarding geofeatures.css, you should only add that to theme.config.php if it is important to load it on every single page of the site. If it is only needed in cases where a map is present, it is better to conditionally load it in the relevant template with `$this->headLink()->appendStylesheet('geofeatures.css');`. My guess is that you would only need to put this in two templates (the MapSelection recommendation module and the Map tab) to get things working... but perhaps I am mistaken.

I hope that is helpful; let me know if you still have questions.

If it's smooth sailing from here, please inform me when you're ready for me to actually test this -- I figure I should wait until you've made any adjustments based on these answers before I take a closer look.

lmgonzales commented:
@demiankatz , thanks for the clarification on those points. 

for 1) One thing we talked about at the summit last year was possibly making it possible to configure where the map selection feature is loaded. So for example, we could have it load in its default location, or perhaps load it as an add-on at the bottom of the advanced search page (like is done here: https://www.finna.fi/Search/Advanced). With that in mind, I have no problem leaving the code as it is to allow for future flexibility. 

for 2) I've removed the changes to the theme.config.php file and inserted the css call in the MapSelection.phtml file. We don't need this css for the Map Tab since it only applies to the clusters. 
 
So at this point, I think it's all ready for you to test. Thanks!

demiankatz commented:
Sounds good -- I'll try to take a look on Monday!

demiankatz commented on config/vufind/config.ini:
```diff
@@ -799,10 +799,11 @@ authors         = Wikipedia
-; (see import/marc_local.properties for more information).
+; These configuration settings have been superseded by the geofeatures.ini file.
+; See the [MapTab] section of the geofeatures.ini file for more information.
+;
```
If geofeatures.ini is the master configuration now, I'd suggest deleting all extra configuration comments that duplicate information found in that file. It makes the documentation easier to maintain if we only have any given piece of it in one place at a time... The "see for more information" note is helpful, but the rest can probably be deleted.

lmgonzales replied:
Makes perfect sense to me. I'll delete all the unnecessary configuration comments and leave the "see for more information..."note in place.

lmgonzales replied:
Should we also remove all of the legacy configuration settings too and only read from the geofeatures.ini file (rather than from config.ini and searches.ini and then geofeatures.ini)? That could streamline the code a lot if we did this.

demiankatz replied:
As long as geofeatures.ini offers the same flexibility as the old code did (and I'm pretty sure it does), I don't mind removing the legacy locations... but if we do, we'll need to be sure to document it, and also add support to VuFind\Config\Upgrader to detect the old settings and move them to the new locations. So ultimately it's a question of whether it's more work to write the upgrade code or retain the legacy compatibility.

demiankatz commented on config/vufind/geofeatures.ini:
```diff
@@ -1,8 +1,15 @@
+[GeoPlatform]
+; Configures the mapping platform that should be used.
+; Replaces config.ini recordMap setting.
+; Options:  leaflet (default), openlayers (legacy)
```
How long do you plan on maintaining openlayers as a legacy platform? Is it worth doing?

lmgonzales replied:
I'd prefer not to maintain it if possible, and honestly, I'm don't think it's really worth it given the improvements with the leaflet code.  I'm happy to make this a clean transition and remove all the OpenLayers code with this PR if you think it's advisable.

demiankatz replied:
If that's the end goal, I think it's probably simpler to just go ahead and do it now.

lmgonzales replied:
Ok. Will do.

demiankatz commented on config/vufind/searches.ini:
```diff
@@ -628,6 +628,9 @@ view=full
 [MapSelection]
+; These configuration settings have been superseded by the geofeatures.ini file.
+; See the [MapSelection] section of the geofeatures.ini file for more information.
+;
```
As noted above, we should delete redundant documentation here. Also note that the [MapSelection] section is referenced a couple other places earlier in this file. We should update those references too if necessary.

lmgonzales replied:
Will do. Thanks for the note.

demiankatz commented on module/VuFind/src/VuFind/Recommend/Factory.php:
```diff
@@ -189,15 +189,18 @@ public static function getFavoriteFacets(ServiceManager $sm)
 
-        return new MapSelection($config, $solr, $basemapOptions);
+        // get MapSelection options
+        $mapSelectionConfig = $sm->get('VuFind\GeoFeatures\MapSelectionConfig');
```
When I first tested this branch, I got a fatal error because there was a `$sm->getServiceLocator()->get()` call here, which no longer works following the upgrade to ServiceManager v3. I've gone ahead and fixed it, but just wanted to let you know. In the old SMv2, factories were passed a plugin-manager, and you had to use getServiceLocator to move up to the parent top-level service manager. In SMv3, the top-level service manager is always the one passed to factories, which makes things a lot more consistent and readable.

lmgonzales replied:
Ok. Good to know and thanks for the catch and explanation.

demiankatz replied:
Fixing this didn't help with the bad request -- it resolved a 500 error that I was getting before I ran into the 400 error. I haven't tested the code since your latest revisions, though; I'm in meetings all day today, and tomorrow we're expecting a blizzard, but I'll try to give it another shot later in the week!

lmgonzales replied:
Ok. Let me see if pulling out all the OpenLayers code and refactoring things fixes this. Let me see if I can reproduce it.

demiankatz replied:
I've mentioned that this problem was fixed elsewhere, but I'm also commenting on it here so we can see that this conversation is resolved when reviewing the PR in future. This is all better now! :-)

demiankatz commented on themes/bootstrap3/templates/RecordTab/map.phtml:
```diff
@@ -1,14 +1,31 @@
+      $popupTitle = $this->transEsc('map_results_label');
+      $params = [json_encode($mapTabData), json_encode($popupTitle), json_encode($mapGraticule), json_encode($basemap)];
+    } elseif ($mapType == 'leaflet') {
+      $mapTabData = $this->tab->getLeafletMapTabData();
```
If you're always setting a $mapTabData variable in the template, I wonder if it would be cleaner to make $this->tab->getMapTabData smarter internally, so it just returns the right kind of data, rather than having to create a separate public getLeafletMapTabData method. That way, if in future you remove the OpenLayers implementation, there's not a Leaflet-specific method name floating around as a potential source of confusion.

(This is a very minor point, by the way -- just something that struck me as a potential code clarification).

lmgonzales replied:
Good point. And it would give us the flexibility to change mapping platforms in the future if something better than leaflet comes along.

demiankatz commented:
@lmgonzales, a couple of new notes:

1.) I had a few minutes before my meeting, so I managed to repeat my tests... I'm still getting the 400 error. The issue seems to be that the geographic search link is pointing to this url: `/Search/Results?filter[]=long_lat:Intersects(ENVELOPE())` -- note the empty ENVELOPE. I assume that means that some configuration is not loading correctly. As noted before, all I did was uncomment the appropriate top recommend setting in searches.ini to turn on geographic search, and it looks like there is a default box set in geofeatures.ini.... so I don't think the configs look wrong by default, but something seems to be preventing them from working.

2.) The intermittent Codacy errors about Javascript indentation are back -- you can see them if you click through the "Details" link on the Codacy review line in the "Some checks were not successful" box. Probably worth cleaning that up if you can.

Let me know if you need more help from my end! Thanks again!

lmgonzales commented:
Thanks for the feedback @demiankatz . I'll work on streamlining the code to remove all the OpenLayers code and this should hopefully fix those issues. I'll also go back and fix the spacing and other things codacy is complaining about with the js files. Thanks again!

demiankatz commented:
@lmgonzales, I've merged the latest master into this branch to save you time on conflict resolution; hopefully I haven't broken anything in the process! Anyway, I imagine you've been busy with other things; no rush to get back to this, especially since I'll be away at a conference most of next week!

lmgonzales commented:
@demiankatz, I've removed all the openlayers code, and found and fixed the issue that was causing the 400 issue. Please let me know if you find any other issues when you get a chance to test the code. Thanks!

demiankatz commented on module/VuFind/src/VuFind/GeoFeatures/MapTabConfig.php:
```diff
@@ -0,0 +1,140 @@
+     *
+     * @return array
+     */
+    protected function getOptions($configName, $section, $validOptions)
```
Since MapSelectionConfig and MapTabConfig both use the same constructor and getOptions function and both have the same dependencies, this is a good candidate for inheritance. I'd recommend creating an AbstractConfig class which contains the configLoader property, constructor and getOptions method. You can then set up MapSelectionConfig and MapTabConfig as subclasses of AbstractConfig. You can then rename either MapSelectionConfigFactory or MapTabConfigFactory to AbstractConfigFactory, and change the configuration so that both classes get built by the same factory. This should accomplish exactly the same behavior with less code.

I seem to recall that your background is not in object oriented PHP, so if you need clarification on any of that, let me know and I'll be happy to elaborate -- this is a good example of why object oriented PHP can be fun, so I hope it's a useful exercise. :-)

lmgonzales replied:
That's a great idea. I will give it a go and if I run into any questions I'll be sure to ask!

demiankatz commented on themes/bootstrap3/templates/Recommend/MapSelection.phtml:
```diff
@@ -1,26 +1,38 @@
-  $params = [json_encode($geoField), json_encode($coordinates), json_encode($urlpath),
-    json_encode($baseUrl), json_encode($searchParams), json_encode($showSelection),
-    json_encode($resultsCoords), json_encode($popupTitle), json_encode($basemap)];
+  $helpTopic = 'geosearch_leaflet';
```
Should we replace the existing help page rather than creating geosearch_leaflet, if we aren't keeping the old functionality anyway?

lmgonzales replied:
Yes, I'll update that.

demiankatz commented on config/vufind/geofeatures.ini:
```diff
@@ -1,8 +1,15 @@
+; Configures the mapping platform that should be used.
+; Replaces config.ini recordMap setting.
+; Options:  leaflet (default)
+geoPlatform = leaflet
```
Do you think it is worth maintaining a geoPlatform setting if we're not worrying about backward compatibility here? It might simplify some things to rip it out. I don't mind keeping it if you think it has value, but if it's unlikely that we're going to offer alternate platforms, it might be just as well to get rid of it, and deal with reimplementing it in the future if the situation changes.

lmgonzales replied:
If we were going to maintain two mapping platforms, then yes, but I'd rather keep this streamlined. I'll need to do some rewriting of the code per this suggestion and the ones above and below, but I think that will be good.

demiankatz commented on config/vufind/geofeatures.ini:
```diff
@@ -40,10 +47,46 @@ basemap_attribution = "<a href=""https://wikimediafoundation.org/wiki/Maps_Terms
+;                coordinate field will be displayed before the map label in the label popup.
+; graticule: true or false. Default is false. If graticule is true a lat/long grid will be
+;                displayed on the map.
+mapLabels        = file:geosearch_test_lookup.txt
```
I assume that we don't really want this setting turned on by default. Do you mind reviewing this file and commenting out anything that we don't actually want to make active (in case there are any other settings that are not intended for default production use)?

lmgonzales replied:
Will do. Thanks

lmgonzales commented on config/vufind/searches.ini:
```diff
@@ -328,13 +328,9 @@ CallNumber = callnumber-sort
-;       section in this file. The section name and ini file name loaded by the
-;       module may be overridden through the [ini section]/[ini name] parameters.
+; MapSelection
+;       These configuration settings have been superseded by the geofeatures.ini file.
```
Thanks Demian. Yes, that's what the notes should read. Thank you. I'm updating the config file with your suggested changes.

lmgonzales commented on module/VuFind/src/VuFind/GeoFeatures/MapSelectionConfig.php:
```diff
@@ -0,0 +1,141 @@
+ *
+ * PHP version 5
+ *
+ * Copyright (C) Villanova University 2010.
```
Sure thing. I've updated all the files in the VuFind/src/VuFind/GeoFeatures directory. I'm happy to update the rest of the geographic feature code too. Please let me know if you would like me to do that.

demiankatz replied:
My global search-and-replace should have taken care of the PHP version change in all existing files, so you should only need to worry about this for newly-added files. Similarly, most of our file copyright dates reflect the date the file was first created, so it might be worth updating these in new files, but you don't have to worry about existing ones (and it's purely cosmetic anyway).

lmgonzales commented on module/VuFind/src/VuFind/GeoFeatures/MapSelectionConfig.php:
```diff
@@ -0,0 +1,140 @@
+    {
+        $config = $this->configLoader->get($configName);
+        $options = [];
+        $fields = $validOptions;
```
Yes, we can definitely do that. Thank you for the recommendation.

lmgonzales commented on module/VuFind/src/VuFind/GeoFeatures/MapSelectionConfig.php:
```diff
@@ -0,0 +1,140 @@
+        $mapPlatform = $this->getOptions('config', 'Content', ['recordMap']);
+
+        // Check geofeatures.ini [MapSelection] section next:
+        if (empty($options)) {
```
Yes, good catch. I'll rewrite this with the other changes suggested below.

lmgonzales commented:
@demiankatz ,

I've updated the code to do the refactoring, bug fixes, and removal of the geoPlatform code as you suggested. Here are some answers to your questions above:

_1.) I notice that when I create a search box, Leaflet produces very visible instructions ("Click and drag to draw rectangle", etc.). This is helpful, but is there the ability to internationalize it? Does Leaflet accept a region code and/or can we pass in translated strings when necessary?_

This is hardcoded into the leaflet.draw.js code. I don't see why we can't modify this file (or make a VuFind version of it) to pass in a variable that reads from the language files for the two strings 'Click and drag to draw rectangle.' and 'Release mouse to finish drawing.'. I might have some questions @crhallberg about how to implement this. 

_2.) For some reason, I'm only seeing clusters in the geographic search, not points. So I can click clusters and watch them expand, but when I click the smallest cluster, it just disappears entirely. There are no markers on the map, and nothing that directs me to a record. I'm seeing correct behavior on your Cold Regions bibliography. Are you seeing the same thing on this branch? Is it possible I just have something misconfigured?_

I think this might be a misconfiguration issue. I'm not seeing this at all. I'd be happy to look at this with you if you have some time to do a quick chat/ screen share. 

_3.) When I zoom in and out on the map tab in Firefox, I repeatedly see TypeError: L.DomUtil.getTranslateString is not a function in my console. I'm also not seeing any markers in this context. Are these things related?_

My guess is that this is related to the issue you're having in point 2 above as I'm not seeing this. 

_4.) Do we have different constraints in Leaflet as far as selecting a large region of the map is concerned, or is that more of a Solr issue? If I want to see all of the points on the map, it's hard for me to make a selection -- if I zoom way out and then create a big box, I tend to end up actually creating a small box (probably because I'm selecting more than the whole world in one go). This may be unavoidable -- but it seems like something that might confuse a naive user._

Perceived constraints may be when a user selects more than one world's extent, and thus only the non-overlapping space is selected. If you stay within one world, or at the least if you don't overlap areas in multiple worlds (so for instance you select from world -right in Asia to world -center in North America), the search box will select as expected.

Please let me know if you're still running into issues.

lmgonzales commented:
@demiankatz and @crhallberg , 
I've started working on implementing a way to internationalize the geographic search tooltip with the latest commit. I am trying to use the VuFind.translate functionality that is present in some of the other .js files in the bootstrap3/js/ directory, but have not been able to get it to work yet. Do I need to register a function in the leaflet_draw_intl.js file to get the VuFind.translate to look up the string in the language/en.ini file? Thanks for your guidance on this.

lmgonzales commented:
@demiankatz, Thanks for the suggestion on how to add support for the internationalization of the tooltips. I've deleted the customized vendor code an inserted a couple of lines into the MapSelection.phtml which will override the tooltips and read from the language .ini files.

demiankatz commented on module/VuFind/src/VuFind/GeoFeatures/BasemapConfig.php:
```diff
@@ -38,22 +36,22 @@
      * @var string
      */
-    protected $basemapUrl = "https://maps.wikimedia.org/osm-intl/{z}/{x}/{y}.png";
+    protected $basemap_url = "https://maps.wikimedia.org/osm-intl/{z}/{x}/{y}.png";
```
Do these basemap_url and basemap_attribution properties serve any purpose? It doesn't look like they're being read anywhere. Also note that for style reasons, if we do need these, it's probably preferable to keep the old camelCase names instead of the under_scored versions... but I'm guessing that we may just be able to remove them entirely.

demiankatz replied:
Oh, wait, I see what's happening -- these are getting dynamically loaded as defaults. I guess I can live with the less-than-ideal naming as a matter of necessity for consistency with the .ini files. The only alternative would be to have a $defaultOptions property that holds all of these values, instead of having a separate property for each configuration.

demiankatz replied:
I suppose yet another possibility would be to have a getDefaultOptions function, so that instead of doing the for loop to populate the options array from class properties, you just have:

<pre>
if (empty($options)) {
    return $this->getDefaultOptions();
}
</pre>

and then elsewhere define:

<pre>
protected function getDefaultOptions()
{
    return [
        'option1' => 'value1',
        'option2' => 'value2';
    ];
}
</pre>

The difference between a defaultOptions property and a getDefaultOptions() function is pretty minimal in most scenarios, but it has some impact on how your code formatting looks (I feel like with a lot of options, the function might look cleaner and make commenting easier) and also how extensibility works (somebody subclassing the code would need to override defaultOptions property details in a custom constructor... while a getDefaultOptions function could be overridden directly).

I don't honestly have a strong opinion about this -- just laying out the options. Sorry if I'm wasting too much time on a trivial matter. :-)

lmgonzales replied:
@demiankatz, 
What if we change the .ini file configuration to use camelCase for the configuration parameters (so we use basemapUrl and basemapAttribution)? It would make the code so that it isn't backwards compatible, but given we are doing a major upgrade, I wonder if that would be okay. What do you think?

Also, I'll take a look into implementing a getDefaultOptions() function tomorrow.

demiankatz replied:
If memory serves, the basemap settings are all new anyway, so changing them won't cause an incompatibility with a previous VuFind release... it will only affect people who are working on bleeding-edge master, of which few if any are using this new geographic feature. In other words, unless my memory is playing tricks, I think it's safe to change this... but you may find that it's unnecessary if getDefaultOptions proves a cleaner solution anyway.

demiankatz commented on module/VuFind/src/VuFind/GeoFeatures/BasemapConfig.php:
```diff
@@ -90,76 +71,61 @@ public function __construct(\VuFind\Config\PluginManager $configLoader)
      */
     public function getBasemap($origin)
     {
+        $validFields = ['basemap_url', 'basemap_attribution'];
```
If $validFields is always going to be the same thing for all getOptions calls made by this class, perhaps it would be better to set it as a property... then you could use $this->validFields in all of the getOptions calls, and avoid the need to pass the value down to getMapSelectionBasemap and getMapTabBasemap. Might improve readability a little bit and also potentially increase extensibility (though I don't think it's incredibly likely that anyone would care about overriding it).

demiankatz commented on themes/root/templates/HelpTranslations/en/geosearch.phtml:
```diff
@@ -11,11 +11,11 @@
-        pointing at the cluster. Click on the cluster, and a pop-up window will appear showing a list of titles for the
-        records at that location. Each title is hyperlinked to the associated record.
-        Click on the title to view the full record.</p>
+        double-click (zoom in) or shift + double-click (zoom out). You can also click on a cluster to zoom to 
```
@xmorave2, the geographic interface is changing slightly due to the switch to Leaflet instead of OpenLayers. Would you have a moment to adjust the Czech translation of this file to match?

lmgonzales commented:
@demiankatz, Thanks for the code improvement suggestions. I've updated the js files so the relative links are gone, and implemented the getDefaultOptions for all of the Config files in the GeoFeatures directory. Please let me know if you see any other improvements/edits/fixes to make. Thanks!

demiankatz commented:
@lmgonzales, thanks again for the latest round of work. This is looking great! Just one small action item and one note:

1.) The text "This is the center point for the highlighted rectangle" is currently hard-coded in map_selection_leaflet.js; we need to run this through the translator. As noted in earlier discussion, you'll probably want to add this string to en.ini, and then use the jsTranslation() view helper to register the translation in the templates that are including map_selection_leaflet.js. Once the string is registered through the template, you can use VuFind.translate() in your Javascript code.

2.) I changed the way that getDefaultOptions() is used in all of your config loading classes. Instead of returning defaults only when all configuration files are empty, I'm instead always returning `$options + $this->getDefaultOptions()` -- this means that any settings that are omitted from the .ini file will be automatically populated with default values. I was encountering some weird errors when I filled in one .ini setting but omitted a different one. Merging in the defaults solved the issues. I think this is a further sign that it was the right choice to go this route!

You might want to look over my last few changes and make sure I haven't broken anything. At this stage, I think the finishing touches on translation are the only thing holding us up from merging.

demiankatz commented:
Thanks again, @lmgonzales! I have just merged this after making a few more small changes.

demiankatz commented:
@lmgonzales, if you think I should add anything more to the [changelog](https://vufind.org/wiki/changelog), please let me know. I put a couple of notes in the 5.0 section.

lmgonzales commented:
Thanks Demian. Yes, updating the changelog and the Geographic Features
pages on the wiki are on my to do list.



Kind regards,
Leila



*From:* Demian Katz [mailto:notifications@github.com]
*Sent:* Monday, April 16, 2018 11:28 AM
*To:* vufind-org/vufind
*Cc:* Leila Gonzales; Mention
*Subject:* Re: [vufind-org/vufind] Geofeatures Leaflet implementation
(#1129)



@lmgonzales <https://github.com/lmgonzales>, if you think I should add
anything more to the changelog <https://vufind.org/wiki/changelog>, please
let me know. I put a couple of notes in the 5.0 section.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/vufind-org/vufind/pull/1129#issuecomment-381703442>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AMc0UWEcYn00d71Wk8Vl_eRj8prM8oY5ks5tpOK_gaJpZM4SZdwg>
.
